File: go/schemify/internal/schema/namespace.go
package schema

import (
	"sort"
	"strings"
)

// IsSystemNamespace reports PostgreSQL built-in namespaces we never manage.
func IsSystemNamespace(name string) bool {
	name = strings.ToLower(strings.TrimSpace(name))
	if name == "" {
		return true
	}
	switch name {
	case "pg_catalog", "information_schema":
		return true
	}
	return strings.HasPrefix(name, "pg_toast") || strings.HasPrefix(name, "pg_")
}

// IsDropSchemaCandidate is true when a surplus namespace in the DB should surface as drop_schema drift.
// public is never a candidate (asymmetric rule: we may create it but never plan-drop it).
func IsDropSchemaCandidate(name string) bool {
	name = strings.ToLower(strings.TrimSpace(name))
	if name == "" || name == "public" {
		return false
	}
	return !IsSystemNamespace(name)
}

// CollectDesiredNamespaces returns all namespace names required by the desired model.
func CollectDesiredNamespaces(load *LoadResult) map[string]struct{} {
	if load == nil {
		return map[string]struct{}{}
	}
	out := make(map[string]struct{})
	add := func(ns string) {
		ns = strings.ToLower(strings.TrimSpace(ns))
		if ns == "" || IsSystemNamespace(ns) {
			return
		}
		out[ns] = struct{}{}
	}
	for ns := range load.Schemas {
		add(ns)
	}
	for _, t := range load.Tables {
		add(t.Schema)
		for _, fk := range t.ForeignKeys {
			ref := fk.ReferencesSchema
			if ref == "" {
				ref = "public"
			}
			add(ref)
		}
	}
	for _, idx := range load.Indexes {
		add(idx.Schema)
		add(idx.TableSchema)
	}
	return out
}

// NamespaceNames returns sorted namespace names from a set.
func NamespaceNames(set map[string]struct{}) []string {
	if len(set) == 0 {
		return nil
	}
	out := make([]string, 0, len(set))
	for ns := range set {
		out = append(out, ns)
	}
	sort.Strings(out)
	return out
}

// UnionNamespaces returns sorted union of two namespace sets.
func UnionNamespaces(a, b map[string]struct{}) []string {
	if len(a) == 0 && len(b) == 0 {
		return nil
	}
	out := make(map[string]struct{})
	for ns := range a {
		out[ns] = struct{}{}
	}
	for ns := range b {
		out[ns] = struct{}{}
	}
	return NamespaceNames(out)
}

File: rust/schemify/src/namespace.rs
//! PostgreSQL namespace (schema) helpers.

use crate::load::LoadResult;
use std::collections::{BTreeSet, HashSet};

pub fn is_system_namespace(name: &str) -> bool {
    let name = name.trim().to_lowercase();
    if name.is_empty() {
        return true;
    }
    match name.as_str() {
        "pg_catalog" | "information_schema" => true,
        _ => name.starts_with("pg_toast") || name.starts_with("pg_"),
    }
}

pub fn is_drop_schema_candidate(name: &str) -> bool {
    let name = name.trim().to_lowercase();
    if name.is_empty() || name == "public" {
        return false;
    }
    !is_system_namespace(&name)
}

pub fn collect_desired_namespaces(load: &LoadResult) -> HashSet<String> {
    let mut out = HashSet::new();
    let mut add = |ns: &str| {
        let ns = ns.trim().to_lowercase();
        if ns.is_empty() || is_system_namespace(&ns) {
            return;
        }
        out.insert(ns);
    };
    for ns in &load.schemas {
        add(ns);
    }
    for t in load.tables.values() {
        add(&t.schema);
        for fk in &t.foreign_keys {
            let ref_schema = if fk.references_schema.is_empty() {
                "public"
            } else {
                fk.references_schema.as_str()
            };
            add(ref_schema);
        }
    }
    for idx in load.indexes.values() {
        add(&idx.schema);
        add(&idx.table_schema);
    }
    out
}

pub fn union_namespaces(a: &HashSet<String>, b: &HashSet<String>) -> Vec<String> {
    let mut out: BTreeSet<String> = BTreeSet::new();
    for ns in a {
        out.insert(ns.clone());
    }
    for ns in b {
        out.insert(ns.clone());
    }
    out.into_iter().collect()
}
